# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.
configfile: "config/config.yaml"
report: "report/workflow.rst"

import pandas as pd

metadata = pd.read_csv(config['samples'], sep="\t")
samples = metadata['Name']

rule all:
    input:
        # The first rule should define the default target files
        # Subsequent target rules can be specified below. They should start with all_*.
        expand("results/quant/{sample}", sample=samples),
	"results/busco/transcriptome.busco.tsv",
	"results/transdecoder/Trinity.fasta.transdecoder.bed",
#        expand("resources/reads/qc/{sample}_{n}_fastqc.html", sample=samples, n=[1,2]),
#        "results/transrate"


#include: "rules/common.smk"
include: "rules/qc.smk"
include: "rules/assembly.smk"
include: "rules/diff.smk"